libraries/AP_Generator/AP_Generator_IE_2400.cpp

```diff
@@ -0,0 +1,217 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "AP_Generator_IE_2400.h"
+
+#if GENERATOR_ENABLED
+
+#include <AP_Logger/AP_Logger.h>
+#include <stdio.h>
+
+// Update fuelcell, expected to be called at 20hz
+void AP_Generator_IE_2400::update()
+{
+    if (_uart == nullptr) {
+        return;
+    }
+
+    const uint32_t now = AP_HAL::millis();
+
+   // Read any available data
+    uint32_t nbytes = MIN(_uart->available(),30u);
```

Comment:
Yeah, thankfully IE were very prescriptive in the data types and size being sent.  It will take the largest string length from a 'valid sentence' from either of the fuel cell units.